src/drivers/smmstore/ramstage.c

```diff
@@ -57,14 +59,24 @@ static void init_store(void *unused)
 
 	printk(BIOS_INFO, "SMMSTORE: Setting up SMI handler\n");
 
-	/* Issue SMI using APM to update the com buffer and to lock the SMMSTORE */
-	__asm__ __volatile__ (
-		"outb %%al, %%dx"
-		: "=a" (eax)
-		: "a" ((SMMSTORE_CMD_INIT << 8) | APM_CNT_SMMSTORE),
-		  "b" (ebx),
-		  "d" (APM_CNT)
-		: "memory");
+	/*
+	 * Issue SMI using APM to update the com buffer and to lock the SMMSTORE.
+	 * Retry 5 times in case the SMI isn't triggered immediately.
+	 */
+	do {
+		__asm__ __volatile__ (
+			"outb %%al, %%dx"
+			: "=a" (eax)
+			: "a" ((SMMSTORE_CMD_INIT << 8) | APM_CNT_SMMSTORE),
+			  "b" (ebx),
+			  "d" (APM_CNT)
+			: "memory");
+
+		if (eax == SMMSTORE_RET_SUCCESS)
+			break;
+
+		mdelay(1);
```

Comment:
@krystian-hebel not sure if retries are good here.

@mkopec have you looked at SMMSTORE_CMD_INIT code? Is it reentrant?

Reply:
This is based on the comment and code in edk2's SMMSTORE v2 driver: https://github.com/MrChromebox/edk2/commit/27854bc8c5e7147f01e50404f6e5b16ffedf08b8#diff-3ce60dbfce1b2c4fb8b00b51d566b6a056a567c819ebedbd4e74fcb0834cdcc9R28-R41

If a previous SMMSTORE_CMD_INIT proceeds successfully but the value of eax isn't successful then the next execution will return SMMSTORE_RET_FAILURE - but assuming the comment in the link above is correct, that will never happen: `[...] in which case the Smi handler will do nothing as only synchronous Smis are handled`

Reply:
https://github.com/Dasharo/coreboot/blob/0c6a6f7d7ac426e0eee5d6ad26846bb68295c6ef/src/drivers/smmstore/store.c#L303 should be safe to call, but would return error anyway. It would be nice to have a way of returning "already initialized" status.